applications/tari_watcher/src/port.rs

```diff
@@ -1,58 +1,76 @@
 // Copyright 2024 The Tari Project
 // SPDX-License-Identifier: BSD-3-Clause
-use std::{
-    io,
-    net::{IpAddr, SocketAddr},
-};
+use std::net::{IpAddr, SocketAddr};
 
+use anyhow::bail;
 use tokio::net::TcpListener;
 
 use crate::config::InstanceType;
 
 #[derive(Clone)]
 pub struct PortAllocator {
     pub validator: ValidatorPorts,
+    pub env: Vec<(String, String)>,
     pub wallet: MinotariPorts,
 }
 
 impl PortAllocator {
-    pub fn new() -> Self {
+    pub fn new(env: Vec<(String, String)>) -> Self {
         Self {
             validator: ValidatorPorts::new(),
+            env,
             wallet: MinotariPorts::new(),
         }
     }
 
-    pub async fn open_at(&mut self, instance: InstanceType, name: &'static str) -> io::Result<u16> {
-        let addr = TcpListener::bind(SocketAddr::from(([127, 0, 0, 1], 0))).await?;
-        let port = addr.local_addr()?.port();
-        match instance {
-            InstanceType::TariValidatorNode => {
-                if name == "jrpc" {
-                    self.validator.jrpc = Some(port)
-                } else if name == "web" {
-                    self.validator.web = Some(port)
-                } else {
-                    log::error!("Invalid port name for {} instance: {}", instance, port);
-                }
-            },
-            InstanceType::MinoTariConsoleWallet => {
-                if name == "p2p" {
-                    self.wallet.p2p = Some(port)
-                } else if name == "grpc" {
-                    self.wallet.grpc = Some(port)
-                } else {
-                    log::error!("Invalid port name for {} instance: {}", instance, port);
-                }
-            },
+    fn vn_json_rpc_port(&self) -> Option<u32> {
```

Comment:
Is there a reason the watcher needs to know the VN's JSON-RPC address?

I don't want the user to have to set some environment vars to setup/start the watcher. 

The watcher doesn't need to talk to the VN's JSON RPC to register it. We only need to read the registration.json file.

The approach I had in mind for determining whether the VN needs to be re-registered was that the watcher periodically polls the L1 minotari node gRPC for the validator node registration (see Registration Management in https://github.com/tari-project/tari-dan/issues/1112). The advantage here is that these calls already exist.

Another approach may be to poll the VN JSON-RPC to check L1 chain progression since VN's does already know this. Then the watcher wouldn't need to talk to the minotari node at all.  Happy to do that, though we'd need to add a new call or modify get_epoch_manager_stats to give the expiry epoch and current block height to allow this. We'd also need to know what port the VN is using for JSON RPC which will be tricky unless we set it in the watcher. 

There are a few ways to do this. Open to thoughts.

Reply:
Likely makes sense to do it with the same pattern for now, unless we otherwise know a clearly better approach